clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -418,10 +419,14 @@ private boolean maybeSendAndPollTransactionalRequest() {
 
         if (transactionManager.hasAbortableError() || transactionManager.isAborting()) {
             if (accumulator.hasIncomplete()) {
+                // Attempt to get the last error that caused this abort.
                 RuntimeException exception = transactionManager.lastError();
+                // If there was no error, but we are still aborting,
+                // then this is most likely a case where there was no fatal error.
                 if (exception == null) {
-                    exception = new KafkaException("Failing batch since transaction was aborted");
+                    exception = new TransactionAbortedException();
                 }
+                // Since the transaction is aborted / being aborted, abort all the undrained batches.
```

Comment:
This comment might be unnecessary, the code is pretty self-explanatory in this case 🙂

Reply:
Aah. Sorry - I'll take that away.